x-pack/solutions/observability/plugins/apm/common/service_map/group_resource_nodes.ts

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { compact, groupBy } from 'lodash';
+import { SPAN_TYPE, SPAN_SUBTYPE } from '../es_fields/apm';
+import type {
+  ConnectionEdge,
+  ConnectionElement,
+  ConnectionNode,
+  GroupResourceNodesResponse,
+  GroupedEdge,
+  GroupedNode,
+} from './types';
+import { getEdgeId, isSpanGroupingSupported } from './utils';
+
+const MINIMUM_GROUP_SIZE = 4;
+
+const isEdge = (el: ConnectionElement): el is { data: ConnectionEdge } =>
+  Boolean(el.data.source && el.data.target);
+const isNode = (el: ConnectionElement): el is { data: ConnectionNode } => !isEdge(el);
+const isElligibleGroupNode = (el: ConnectionElement): el is { data: ConnectionNode } => {
+  if (isNode(el) && SPAN_TYPE in el.data) {
+    return isSpanGroupingSupported(el.data[SPAN_TYPE], el.data[SPAN_SUBTYPE]);
+  }
+  return false;
+};
+
+function groupConnections({
```

Comment:
This method is pretty complex and heavy, with multiple iteration passes. I wonder if we can condense the iteration passes to be as few as possible by reducing the amount of collection conversions we are doing. Once we have our `groupedTargets`, we could perhaps group it directly into a `Map` instead of an `Object`, and iterate and reduce through the `Map`. This should reduce needing to allocate two extra arrays + object, and also remove the need for `groupBy` method from `lodash`.

Reply:
Yeah. I think we could reduce the number of iterations here.